### PR TITLE
http: making upstream ALPN accesible.

### PR DESCRIPTION
**Pull request by alyssawilk**
Part of https://github.com/envoyproxy/envoy-mobile/issues/1546

Risk Level: Low
Testing: new integration test
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -1406,14 +1406,17 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
 
   downstream_response_started_ = true;
   final_upstream_request_ = &upstream_request;
-  // In upstream request hedging scenarios the upstream connection ID set in onPoolReady might not
-  // be the connection ID of the upstream connection that ended up receiving upstream headers. Thus
-  // reset the upstream connection ID here with the ID of the connection that ultimately was the
-  // transport for the final upstream request.
+  // In upstream request hedging scenarios properties set in onPoolReady might not be the same
```
Comment (RyanTheOptimist) [inline source/common/router/router.cc:1409]: nit "scenarios" => "scenarios,"

File: source/common/router/router.cc [hunk 2]
```diff
@@ -1406,14 +1406,17 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
...
+  // as the properties set on the final tream, thus we reset fields based on the
+  // final upstream request.
```
Comment (RyanTheOptimist) [inline source/common/router/router.cc:1411]: Wow, TIL: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_routing#arch-overview-http-routing-hedging

Does this mean that the request is actually sent multiple times? (Not just that multiple streams were requested?) I think that's what the hedging policy docs say. Neat! Do you know what this means, "The implementation ensures that the same upstream request is not retried twice."? Is that saying that we won't issue the request twice on the same upstream connection?

**Review by RyanTheOptimist — APPROVED** [review]
Looks good to me. I totally agree that this "incrementally adding upstreamFoo" approach is annoying and for Envoy Mobile, it's likely to be really annoying. I tend to agree that having access to the StreamInfo from the upstream stream is a better path forward.

> cc @RyanTheOptimist I'm also planning on faking "negotiated ALPN" from upstream protocol (so if you negotiate something non-standard it'd not be reported. LMK if you think that's a problem
 
 I think this sounds fine for now, but we should probably wire up the actual TLS ALPN at some point.

**Comment by alyssawilk** [comment]
Ah if we want real ALPN I might as well do that now
/wait

**Comment by alyssawilk** [comment]
OK, switching this over to real ALPN didn't work becasue of a series of refactors over in https://github.com/envoyproxy/envoy/issues/17168 where we made the socket the entity for a bunch of non-address things like the ssl info and the connection id.

This may work just fine for HTTP/1 and HTTP/2 but I'm not sure it's the correct API choice for HTTP/3 support.  I'm working around it for now with a shim and a TODO but I find the shim pretty ugly and I'm wondering if we need to look at changing the APIs again.
cc @ggreenway

**Comment by alyssawilk** [comment]
Arguably we would have had to shim anyway for HTTP/3 since with we may have one socket and many addresses, so maybe it's not actually more problematic?  cc @danzh2010 and @RyanTheOptimist for thoughts as well.
Dan as TODO'd I think we need to look at all these and make sure they work as intended.

**Comment by mattklein123** [comment]
At a high level these seems reasonable to me. Happy to figure out better APIs also if needed. For posterity the other API changes were done to fix some fairly serious state bugs in which we didn't have state tracked correctly across the connection, stream info, etc.

**Review by RyanTheOptimist — APPROVED** [review]
Other than some annoyance at classes outside the scope of this PR, this looks good to me. I think the shim seems totally reasonable to me. We're going to need *something* along these lines and this seems fine.

File: source/common/quic/quic_filter_manager_connection_impl.h [hunk 3]
```diff
@@ -31,7 +31,8 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase,
 public:
   QuicFilterManagerConnectionImpl(QuicNetworkConnection& connection,
                                   const quic::QuicConnectionId& connection_id,
-                                  Event::Dispatcher& dispatcher, uint32_t send_buffer_limit);
+                                  Event::Dispatcher& dispatcher, uint32_t send_buffer_limit,
+                                  std::shared_ptr<QuicSslConnectionInfo> info);
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/quic/quic_filter_manager_connection_impl.h:35]: Shouldn't it be std::shared_ptr<QuicSslConnectionInfo>&& instead?

File: source/common/quic/quic_filter_manager_connection_impl.h [hunk 4]
```diff
@@ -59,11 +60,63 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase,
   void readDisable(bool /*disable*/) override { ASSERT(false); }
   void detectEarlyCloseWhenReadDisabled(bool /*value*/) override { ASSERT(false); }
   bool readEnabled() const override { return true; }
+
+  // TODO(alyssawilk, danzh), sort out which of these need to be handled
+  // locally.
+  class ConnectionInfoProviderShim : public Network::ConnectionInfoSetter {
```
Comment (danzh2010) [inline source/common/quic/quic_filter_manager_connection_impl.h:66]: Sorry, I wasn't able to figure out why we need this shim. Doesn't `network_connection_->connectionSocket()->connectionInfoProvider().setSslConnection(quic_ssl_info_)` work? If not, a comment about this wrapper class will be very appreciated.

File: test/integration/multiplexed_upstream_integration_test.cc [hunk 5]
```diff
@@ -111,6 +111,11 @@ TEST_P(MultiplexedUpstreamIntegrationTest, TestSchemeAndXFP) {
 
 // Ensure Envoy handles streaming requests and responses simultaneously.
 void MultiplexedUpstreamIntegrationTest::bidirectionalStreaming(uint32_t bytes) {
+  config_helper_.prependFilter(fmt::format(R"EOF(
+  name: stream-info-to-headers-filter
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply test/integration/multiplexed_upstream_integration_test.cc:115]: Correct, the test l7 filter is static linked in via RegisterFactory.

File: source/extensions/transport_sockets/tls/connection_info_impl_base.cc [hunk 6]
```diff
@@ -190,6 +190,18 @@ const std::string& ConnectionInfoImplBase::tlsVersion() const {
   return cached_tls_version_;
 }
 
+const std::string& ConnectionInfoImplBase::alpn() const {
+  if (alpn_.empty()) {
```
Comment (danzh2010) [inline source/extensions/transport_sockets/tls/connection_info_impl_base.cc:194]: As we are hard-coding "h3" in server session right now. Would this cause inconsistency between the ALPN returned via Ssl::ConnectionInfo and the one returned via ConnectionSocket::requestedApplicationProtocols()?  requestedApplicationProtocols() is only used in server side, so I guess the upstream connection's connection socket doesn't have ALPN set there, either will it be used. But the changes in this PR applies to downstream as well, hence HCM or the filters can use both the new interface and requestedApplicationProtocols() to get the downstream ALPN and they can be different. Maybe overriding this implementation in QuicSslConnectionInfo according to perspective?

File: source/common/quic/quic_filter_manager_connection_impl.h (hunk 4 above)
Comment (alyssawilk) [reply source/common/quic/quic_filter_manager_connection_impl.h]: ah, I thought quic connections shared the socket object along with the kernel socket.  I feel much better about this PR if I can get rid of this mess.

File: source/extensions/transport_sockets/tls/connection_info_impl_base.cc (hunk 6 above)
Comment (alyssawilk) [reply source/extensions/transport_sockets/tls/connection_info_impl_base.cc:194]: aren't we hard-coding h3 as the thing to negotiate?  I think returning what boring returns is correct here.

**Comment by alyssawilk** [comment]
my new assert broke a unit test feel free to 
/wait
 for the fix

**Review by RyanTheOptimist — COMMENTED** [review]
Nice, much cleaner! Other than the question about "h3" hardcoding, this LGTM. (I wonder what it would take to stop hard coding "h3" in server session)

**Comment by danzh2010** [comment]
> Nice, much cleaner! Other than the question about "h3" hardcoding, this LGTM. (I wonder what it would take to stop hard coding "h3" in server session)

It needs a QUICHE change: https://github.com/envoyproxy/envoy/issues/18935

File: source/extensions/transport_sockets/tls/connection_info_impl_base.cc [hunk 7]
```diff
@@ -190,6 +190,18 @@ const std::string& ConnectionInfoImplBase::tlsVersion() const {
...
+    const unsigned char* proto;
+    unsigned int proto_len;
+    SSL_get0_alpn_selected(ssl(), &proto, &proto_len);
+    if (proto != nullptr) {
+      alpn_ = std::string(reinterpret_cast<const char*>(proto), proto_len);
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/transport_sockets/tls/connection_info_impl_base.cc:199]: Sorry quick question: this code must exist somewhere else, right? Like in TLS inspector or somewhere else? Should we factor it out into some utility function?

/wait-any

**Review by RyanTheOptimist — APPROVED** [review]
LGTM, redux!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**